File: Donut/Donut/Donut/SubScreen.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Charlotte.Tools;
using DxLibDLL;

namespace Charlotte.Donut
{
	public class SubScreen : IDisposable
	{
		private static List<SubScreen> SubScreenList = new List<SubScreen>();

		public int W;
		public int H;
		public bool AFlag;
		public int Handle;

		public SubScreen(int w, int h, bool aFlag = false) // Codevil の CreateSubScreen()
		{
			if (
				w < 1 || IntTools.IMAX < w ||
				h < 1 || IntTools.IMAX < h
				)
				throw new GameError();

			this.W = w;
			this.H = h;
			this.AFlag = aFlag;
			this.Handle = -1;

			SubScreenList.Add(this);
		}

		public void Dispose() // Codevil の ReleaseSubScreen()
		{
			throw new NotImplementedException();
		}

		public int GetHandle()
		{
			if (this.Handle == -1)
			{
				this.Handle = DX.MakeScreen(this.W, this.H, this.AFlag ? 1 : 0);

				if (this.Handle == -1)
					throw new GameError();
			}
			return this.Handle;
		}

		public I2Size GetSize() // Codevil の GetSubScreenSize()
		{
			return new I2Size()
			{
				W = this.W,
				H = this.H,
			};
		}

		// ここから static

		public static int CurrDrawScreenHandle = DX.DX_SCREEN_BACK;

		public static void ChangeDrawScreen(int handle)
		{
			if (DX.SetDrawScreen(handle) != 0)
				throw new GameError();

			CurrDrawScreenHandle = handle;
		}

		public static void ChangeDrawScreen(SubScreen screen)
		{
			ChangeDrawScreen(screen.GetHandle());
		}

		public static void RestoreDrawScreen()
		{
			ChangeDrawScreen(GameGround.I.MainScreen != null ? GameGround.I.MainScreen.GetHandle() : DX.DX_SCREEN_BACK);
		}

		public static void UnloadAllSubScreenHandle()
		{
			ChangeDrawScreen(DX.DX_SCREEN_BACK); // これから開放するハンドルであるとマズいので...

			for (int index = 0; index < SubScreenList.Count; index++)
			{
				SubScreen i = SubScreenList[index];

				if (i.Handle != -1)
				{
					if (DX.DeleteGraph(i.Handle) != 0) // ? 失敗
						throw new GameError();

					i.Handle = -1;
				}
			}
		}
	}
}

File: wb/t20190402/t0001/t0001/t0001/Common/SubScreen.cpp
/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
#include "all.h"

/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
static oneObject(autoList<SubScreen_t *>, new autoList<SubScreen_t *>(), GetSubScreenList);

/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
SubScreen_t *CreateSubScreen(int w, int h, int aFlag)
{
	errorCase(w < 1 || IMAX < w);
	errorCase(h < 1 || IMAX < h);

	SubScreen_t *i = nb_(SubScreen_t);

	i->W = w;
	i->H = h;
	i->AFlag = aFlag;
	i->Handle = -1;

	GetSubScreenList()->AddElement(i);
	return i;
}
/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
void ReleaseSubScreen(SubScreen_t *i)
{
	if(i == NULL)
		return;

	for(int index = 0; index < GetSubScreenList()->GetCount(); index++)
	{
		if(GetSubScreenList()->GetElement(index) == i)
		{
			GetSubScreenList()->FastDesertElement(index);
			goto endDesert;
		}
	}
	error(); // not found
endDesert:

	if(i->Handle != -1)
		errorCase(DeleteGraph(i->Handle)); // ? 失敗

	memFree(i);
}

// <-- cdtor

/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
int GetHandle(SubScreen_t *i)
{
	errorCase(i == NULL);

	if(i->Handle == -1)
	{
		i->Handle = MakeScreen(i->W, i->H, i->AFlag ? TRUE : FALSE);
		errorCase(i->Handle == -1); // ? 失敗
	}
	return i->Handle;
}

// <-- accessor

/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
int CurrDrawScreenHandle = DX_SCREEN_BACK;

/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
void ChangeDrawScreen(int handle)
{
	errorCase(SetDrawScreen(handle)); // ? 失敗
	CurrDrawScreenHandle = handle;
}
/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
void ChangeDrawScreen(SubScreen_t *screen)
{
	ChangeDrawScreen(GetHandle(screen));
}
/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
void RestoreDrawScreen(void)
{
	ChangeDrawScreen(Gnd.MainScreen ? GetHandle(Gnd.MainScreen) : DX_SCREEN_BACK);
}

/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
i2D_t GetSubScreenSize(SubScreen_t *screen)
{
	i2D_t size;

	size.X = screen->W;
	size.Y = screen->H;

	return size;
}
/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
i2D_t GetDrawScreenSize(void) // 描画領域のサイズ？
{
	int w;
	int h;
	int cbd;

	errorCase(GetScreenState(&w, &h, &cbd));
	errorCase(!m_isRange(w, 1, IMAX));
	errorCase(!m_isRange(h, 1, IMAX));

	i2D_t size;

	size.X = w;
	size.Y = h;

	return size;
}
/*
	copied the source file by https://github.com/stackprobe/Factory/blob/master/SubTools/CopyLib.c
*/
void UnloadAllSubScreenHandle(void)
{
	ChangeDrawScreen(DX_SCREEN_BACK); // これから開放するハンドルであるとマズいので...

	for(int index = 0; index < GetSubScreenList()->GetCount(); index++)
	{
		SubScreen_t *i = GetSubScreenList()->GetElement(index);

		if(i->Handle != -1)
		{
			errorCase(DeleteGraph(i->Handle)); // ? 失敗
			i->Handle = -1;
		}
	}
}
